### ISSUE TITLE
coap-client always uses same transaction id

### ISSUE THREAD

**Issue by BlueBeN82** [obgm/libcoap#373]
When I try to send a CoAP message via coap-client it uses the same transaction id (tid) for every call. This leads to every n+1 messages being ignored by the server.

```
bash-4.4# echo '{"some": "json"}' | coap-client -m put -b 1024 -v 9 coap://1.2.3.4:5683/2/1
Jul 03 07:23:43 DEBG ***10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : new outgoing session
Jul 03 07:23:43 DEBG sending CoAP request:
Jul 03 07:23:43 DEBG *  10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:23:43 DEBG ** 10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : tid=47399 added to retransmit queue (2656ms)
Jul 03 07:23:43 DEBG timeout is set to 90 seconds
Jul 03 07:23:46 DEBG ** 10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : tid=47399: retransmission #1
Jul 03 07:23:46 DEBG *  10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
^CJul 03 07:23:50 DEBG ***10.42.91.73:56136 <-> 1.2.3.4:5683 UDP : session closed


bash-4.4# echo '{"some":{"other_json"}' | coap-client -m put -b 1024 -v 7 coap://1.2.3.4:5683/2/1
Jul 03 07:46:02 DEBG ***10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : new outgoing session
Jul 03 07:46:02 DEBG *  10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:46:02 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399 added to retransmit queue (2656ms)
Jul 03 07:46:02 DEBG timeout is set to 90 seconds
Jul 03 07:46:05 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399: retransmission #1
Jul 03 07:46:05 DEBG *  10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:46:11 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399: retransmission #2
Jul 03 07:46:11 DEBG *  10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:46:22 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399: retransmission #3
Jul 03 07:46:22 DEBG *  10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:46:44 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399: retransmission #4
Jul 03 07:46:44 DEBG *  10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : sent 14 bytes
v:1 t:CON c:PUT i:b927 {} [ Uri-Port:4096, Uri-Path:2, Uri-Path:1, Block1:0/_/1024 ]
Jul 03 07:47:27 DEBG ** 10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : tid=47399: give up after 4 attempts
Jul 03 07:47:32 INFO timeout
Jul 03 07:47:32 DEBG ***10.42.91.73:52526 <-> 1.2.3.4:5683 UDP : session closed
```

As you can see in this example it's using the same tid 47399 (0xb927) for two subsequent calls.
If I understood well this should not happen. Any ideas on that?

Context:
- Using alpine linux 3.9.4
- coap-client v4.2.0
- TLS Library: OpenSSL - runtime 1.1.1b, libcoap built for 1.1.1b

**Comment by mrdeep1** [comment]
It appears that prng() always generates the same value on startup of the coap-client as prng() is initially seeded by pring_init(0)  (these 2 functions are equivalent to rand() and srand() I am guessing on your host system).

Strictly speaking the message id (transaction id) has to be unique when sending between 2 endpoints - and the endpoint definition includes the port.

Individual blocks within a bulk data transfer need their own message ID, but as the packet is being repeated in your example, it should have the same message id.

However using the same message id on a restart of a coap-client can be confusing and should be looked into.

**Comment by BlueBeN82** [comment]
Hi! Thx for the quick reply. Any idea what i can do to workaround that issue? 
Can I help somehow on investigating it (without C knowledge)?

**Comment by mrdeep1** [comment]
As there are no response packets from the server, I would say that the server is not up, or is unreachable from your client which is your real issue - not the message id.

**Comment by BlueBeN82** [comment]
Yes and no... the server/device is reachable with another coap library. Also the very first try with coap-client is working as expected. But each subsequent call is not. The issue behind that is, that the coap library used on the server side only looks for the message id to decide wether the message is a dublicate or not.

So in order to be able to use coap-client with the current server/device implementation, there will be no other option rather than using individual tid/mid per message. Same tid for retransmission is perfectly okay.

Is there a way to for the client to use another prng implementation?

**Comment by BlueBeN82** [comment]
The tid are as expected with ubuntu and 4.1.2. Trying with 4.2.0 now...

**Comment by mrdeep1** [comment]
You can try making this change in the code and re-compiling the libcoap library
````
diff --git a/src/net.c b/src/net.c
index 82d222b..469ef68 100644
--- a/src/net.c
+++ b/src/net.c
@@ -2385,7 +2385,7 @@ void coap_startup(void) {
 #elif defined(WITH_CONTIKI)
   prng_init(0);
 #elif !defined(_WIN32)
-  prng_init(0);
+  prng_init(time(NULL));
 #endif
   coap_dtls_startup();
 }
````

**Comment by mrdeep1** [comment]
Yes - commit 7fb4ace2b709994bbbebab4c2d367ac76140eee9 effectively changed prng_init() from having a random parameter to having a parameter value of 0.  I will work on a fix for this.

**Comment by BlueBeN82** [comment]
Glad to hear you found the issue and will fix it! Thanks!
Meanwhile i tried with your changes, but without success.
This is what I've done:

``` 
git clone https://github.com/obgm/libcoap.git
cd libcoap
changed the file as described
./autogen.sh
./configure --disable-doxygen --disable-manpages
make
make install
``` 
As far as I can see there where no errors

coap-client points to the right binary
```
bash-4.4# which coap-client
/usr/local/bin/coap-client
```

**Comment by mrdeep1** [comment]
That should have worked.  You may need to run /sbin/ldconfig to make sure the updated libraries are picked up.

Check that the coap-client and libcoap libraries under /usr/local got updated - and that there are no libcoap libraries lurking under some other directory.

I get (as expected) different i: values each time I run it.
````
$ coap-client -v9 coap://192.168.0.189
Jul 03 13:12:30 DEBG ***192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : new outgoing session
Jul 03 13:12:30 DEBG sending CoAP request:
Jul 03 13:12:30 DEBG *  192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : sent 4 bytes
v:1 t:CON c:GET i:4a21 {} [ ]
Jul 03 13:12:30 DEBG ** 192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : tid=18977 added to retransmit queue (2688ms)
Jul 03 13:12:30 DEBG timeout is set to 90 seconds
Jul 03 13:12:30 WARN coap_network_read: unreachable
Jul 03 13:12:30 DEBG ***192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : session disconnected (reason 2)
Jul 03 13:12:30 DEBG ** 192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : tid=18977: removed
Jul 03 13:12:30 DEBG ***192.168.0.189:53448 <-> 192.168.0.189:5683 UDP : session closed
fix_prng_init [root@ds_192_168_0_189 /Misc/github/libcoap]
$ coap-client -v9 coap://192.168.0.189
Jul 03 13:12:32 DEBG ***192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : new outgoing session
Jul 03 13:12:32 DEBG sending CoAP request:
Jul 03 13:12:32 DEBG *  192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : sent 4 bytes
v:1 t:CON c:GET i:5bd2 {} [ ]
Jul 03 13:12:32 DEBG ** 192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : tid=23506 added to retransmit queue (2406ms)
Jul 03 13:12:32 DEBG timeout is set to 90 seconds
Jul 03 13:12:32 WARN coap_network_read: unreachable
Jul 03 13:12:32 DEBG ***192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : session disconnected (reason 2)
Jul 03 13:12:32 DEBG ** 192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : tid=23506: removed
Jul 03 13:12:32 DEBG ***192.168.0.189:40196 <-> 192.168.0.189:5683 UDP : session closed
````

**Comment by BlueBeN82** [comment]
I've now executed `/sbin/ldconfig`, manually deleted coap-client in /usr/local/bin and libcoap* in /usr/local/lib. The did the whole autogen...make install process again. 
Sadly, I had no success...
May there be some library missing in my alpine linux (which runs inside a docker container)?

**Comment by mrdeep1** [comment]
Hmm - if the clients are started within a second of each other, then the i: will remain the same with the fix I gave you.

Otherwise, the time() implementation may always give you the same value (unlikely as time should increment!) or the rand() implementation is not properly supported in the library and is not generating random numbers, or the srand() function does nothing.

However, I think I read correctly that you said that it was working with the 4.1.2 library, so srand() and rand() should be working fine.

**Comment by BlueBeN82** [comment]
Okay... got it working now. Had to delete the /usr/local/includes/coap* files as well.

**Comment by mrdeep1** [comment]
Good news.

**Comment by BlueBeN82** [comment]
Hi! Just continued to test and in my scenario millisecond is not enough...

```
bash-4.4# cat test.sh
#!/usr/bin/env bash
echo '{"some":"json"}' | coap-client -m put -v 9 -b 1024 coap://1.2.3.4:5683/8/1 >> ~/test &
echo '{"some":"other_json"}' | coap-client -m put -v 9 -b 1024 coap://1.2.3.4:5683/8/2 >> ~/test &
```
With the following script I can send two messages in parallel, which will use the same tid/mid.

Would it be possible to use nanosecond precision or alternatively use milliseconds with e.g. the pid of the current process?

Thanks for being patient ;-)

**Comment by mrdeep1** [comment]
Are you using the code changes from both PR #374 and PR #375?

**Comment by mrdeep1** [comment]
````
diff --git a/src/net.c b/src/net.c
index 152bcaa..8fedaa4 100644
--- a/src/net.c
+++ b/src/net.c
@@ -2358,6 +2358,8 @@ coap_can_exit(coap_context_t *context) {
 static int coap_started = 0;
 
 void coap_startup(void) {
+  coap_tick_t now;
+  coap_time_t t;
   if (coap_started)
     return;
   coap_started = 1;
@@ -2367,13 +2369,10 @@ void coap_startup(void) {
   WSAStartup(wVersionRequested, &wsaData);
 #endif
   coap_clock_init();
-#if defined(WITH_LWIP)
-  prng_init(LWIP_RAND());
-#elif defined(WITH_CONTIKI)
-  prng_init(0);
-#elif !defined(_WIN32)
-  prng_init(0);
-#endif
+  coap_ticks(&now);
+  t = coap_ticks_to_rt(now);
+  /* Be accurate to the nearest (approx) ms */
+  prng_init((t << 10) + (now &0x3ff));
   coap_dtls_startup();
 }
````
````
diff --git a/src/coap_time.c b/src/coap_time.c
index 0931b00..0193319 100644
--- a/src/coap_time.c
+++ b/src/coap_time.c
@@ -68,13 +68,13 @@ coap_clock_init(void) {
 }
 
 /* creates a Qx.frac from fval */
-#define Q(frac,fval) ((coap_tick_t)(((1 << (frac)) * (fval))))
+#define Q(frac,fval) ((1 << (frac)) * (fval))
 
 /* number of frac bits for sub-seconds */
 #define FRAC 10
 
 /* rounds val up and right shifts by frac positions */
-#define SHR_FP(val,frac) (((val) + (1 << ((frac) - 1))) >> (frac))
+#define SHR_FP(val,frac) (((coap_tick_t)((val) + (1 << ((frac) - 1)))) >> (frac))
 
 void
 coap_ticks(coap_tick_t *t) {
````

**Comment by BlueBeN82** [comment]
No, currently i'm only using https://github.com/obgm/libcoap/pull/374

**Comment by mrdeep1** [comment]
Then we currently only have a granularity of a second until #375 is included.  The change is moving `(coap_tick_t)` out of `Q` and put into `SHR_FP` in src/coap_time.c

**Comment by BlueBeN82** [comment]
seems to be working :+1:

**Comment by mrdeep1** [comment]
There is an outside chance that there could be a collision, but unlikely.

**Comment by BlueBeN82** [comment]
Hi! Sorry to say, but we're still hitting the collisions. Wouldn't it be easier and safer to use the current process pid? Or event more secure current implementation plus pid? Each command line call will have a seperate pid (at least per host/container. Another option would be adding the numeric version of the ipv4/ipv6 address of the host system as well. What do you think?

**Comment by obgm** [comment]
Sorry, but the issue you have is clearly on the server side. The message id needs to be unique only in combination with the client's transport address (because the client has no control over the message ids that are generated by different clients).

**Comment by mrdeep1** [comment]
libcoap is supported in Windows.  To get the pid, the call is getpid().  Visual Studio states this

> This API cannot be used in applications that execute in the Windows Runtime.

so, the pid is not necessarily a way to go - and again there is the possibility of collisions from multiple clients in different containers.

#374 was updated today to use micro-sec resolution - #375 has now been merged into develop.  Worth trying out this update.

**Comment by BlueBeN82** [comment]
Okay! Will try with the microsecond precision 👍

**Comment by BlueBeN82** [comment]
Hi @mrdeep1!

I've now tried the microsecond precision, but without success again. We still have collisions.
We are running libcoap with multiple parallel executions from RabbitMQ with PHP worker. So I guess the problem will persist with any time precision. There will always be situations where two process will run at the exact same time. 

Would it be possible to use something like the currently allocated memory or other system resources as additional seed?

**Comment by BlueBeN82** [comment]
Maybe this link is helpful as well?!
https://stackoverflow.com/questions/822323/how-to-generate-a-random-int-in-c

**Comment by mrdeep1** [comment]
Whatever you choose as your initial random seed, there is always a chance of a collision.

As obgm states, the real issue is with your server and that needs to be fixed.  Using message ID on its own to detect duplicate sessions runs a high risk of collisions from different clients and hence rejection - what you are seeing.

RFC7252 requires that the message ID is only unique on a per endpoint basis, and the endpoint definition comprises of both the IP and the port number.  The server, on a IP/port basis should only check for message ID duplication.

I do appreciate that a number of rogue clients launching an attack against a server, all using the same message ID can be picked up by your server's current mitigation technique - but the attack simply then needs to morph into random message IDs and then get through.

Bottom line is the server needs to be fixed - not applying more band-aid to the client logic.

**Comment by mrdeep1** [comment]
@BlueBeN82  Can this be closed off now?

**Comment by BlueBeN82** [comment]
Yes! Your explaination is right and we have to adapt. 
Thanks :-)
